audioswitch/src/main/java/com/twilio/audioswitch/AudioDeviceSelector.kt

```diff
@@ -0,0 +1,368 @@
+    // Saved Audio Settings
+    private var savedAudioMode = 0
+
+    private var savedIsMicrophoneMuted = false
+    private var savedSpeakerphoneEnabled = false
+    private enum class State {
+        STARTED, ACTIVATED, STOPPED
+    }
+    private val EARPIECE_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.EARPIECE, "Earpiece")
+
+    private val SPEAKERPHONE_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.SPEAKERPHONE, "Speakerphone")
+    private val WIRED_HEADSET_AUDIO_DEVICE = AudioDevice(AudioDevice.Type.WIRED_HEADSET, "Wired Headset")
+    private var bluetoothAudioDevice: AudioDevice? = null
+    private val bluetoothController: BluetoothController? = BluetoothController.newInstance(
+            context,
+            LogWrapper(),
+        object : BluetoothDeviceConnectionListener {
+            override fun onBluetoothConnected(
+                bluetoothDevice: BluetoothDevice
+            ) {
+                bluetoothAudioDevice = AudioDevice(
+                        AudioDevice.Type.BLUETOOTH,
+                        bluetoothDevice.name)
+                if (state == State.ACTIVATED) {
+                    userSelectedDevice = bluetoothAudioDevice
+                }
+                enumerateDevices()
+            }
+
+            override fun onBluetoothDisconnected() {
+                bluetoothAudioDevice = null
+                enumerateDevices()
+            }
+        }
+    )
+    internal constructor(context: Context, bluetoothController: BluetoothController) : this(context) {
+    }
+
+    /**
+     * Starts listening for audio device changes. **Note:** When audio device listening is no
+     * longer needed, [AudioDeviceSelector.stop] should be called in order to prevent a
+     * memory leak.
+     *
+     * @param listener receives audio device change events
+     */
+    fun start(listener: AudioDeviceChangeListener) {
+        audioDeviceChangeListener = listener
+        when (state) {
+            STOPPED -> {
+                bluetoothController?.start()
+                context.registerReceiver(
+                        wiredHeadsetReceiver, IntentFilter(Intent.ACTION_HEADSET_PLUG))
+                /*
+                 * Enumerate devices when the wired headset receiver does not broadcast an action.
+                 * The broadcast receiver will not broadcast an action when a wired headset is not
+                 * initially plugged in. This task is intentionally run immediately after the wired
+                 * headset broadcast receiver task runs.
+                 */
+                val handler = Handler()
+                handler.post {
+                    if (!wiredHeadsetAvailable) {
+                        enumerateDevices()
+                    }
+                }
+                state = State.STARTED
+            }
+            State.STARTED, State.ACTIVATED -> {
+            }
+        }
+    }
+
+    fun blah(someInt: SomeInterface) {}
```

Comment:
nix

Reply:
Oh shoot I forgot to pull out that test interface. Thanks!